File: 72_Queen/perl/queen.pl
#!/usr/bin/env perl
use v5.24;
use warnings;
use experimental 'signatures';
no warnings 'experimental::signatures';

use constant TARGET => 158;

main(@ARGV);

sub main (@args) {
   welcome();
   help() if ask_yes_no('DO YOU WANT INSTRUCTIONS');
   do { one_match() } while ask_yes_no('ANYONE ELSE CARE TO TRY');
   __exit();
}

sub one_match {
   print_board();

   # the player can choose the starting position in the top row or the
   # right column
   my $move = ask_first_move() or return forfeit();

   # we alternate moves between computer or player from now on
   while ('playing') {
      $move = computer_move($move);
      say "COMPUTER MOVES TO SQUARE $move";
      return print_computer_victory() if $move == TARGET;

      $move = ask_player_move($move) or return forfeit();
      return print_player_victory() if $move == TARGET;
   }
}

sub is_valid_move ($move, $current, $skip_prevalidation = 0) {

   # pre-validation is needed for moves coming from the user
   if (! $skip_prevalidation) {
      state $valid_position = { map { $_ => 1 } board_identifiers() };
      return 0 unless $move =~ m{\A [1-9]\d+ \z}mxs;
      return 1 if $move == 0;
      return 0 unless $valid_position->{$move};
      return 0 if $move <= $current;
   }

   # the move might be valid in general, let's check from $current
   my $delta = $move - $current;

   # a valid move differs from the current position by a multiple of 10,
   # or 11, or 21. If dividing by all of them yields a remainder, then
   # the move is not valid
   return 0 if $delta % 10 && $delta % 11 && $delta % 21;

   # otherwise it is
   return 1;
}

sub ask_player_move ($current) {
   while ('necessary') {
      my $move = ask_input('WHAT IS YOUR MOVE');
      return $move if is_valid_move($move, $current);
      say "\nY O U   C H E A T . . .  TRY AGAIN";
   }
}

sub computer_move ($current) {

   # this game has some optimal/safe positions from where it's possible
   # to win with the right strategy. We will aim for them, if possible
   state $optimals = [ 158, 127, 126, 75, 73 ];
   for my $optimal ($optimals->@*) {

      # moves can only increase, if we did not find any optimal move so far
      # then there's no point going on
      last if $optimal <= $current;

      # computer moves are "syntactically" valid, skip pre-validation
      return $optimal if is_valid_move($optimal, $current, 'skip');

   }

   # cannot reach an optimal position... resort to randomness
   my $z = rand();
   return $current + 11 if $z > 0.6; # move down
   return $current + 21 if $z > 0.3; # move diagonally
   return $current + 10;           ; # move horizontally
}

sub board_identifiers {
   return (
      81,   71,  61,  51,  41,  31,  21,  11,
      92,   82,  72,  62,  52,  42,  32,  22,
      103,  93,  83,  73,  63,  53,  43,  33,
      114, 104,  94,  84,  74,  64,  54,  44,
      125, 115, 105,  95,  85,  75,  65,  55,
      136, 126, 116, 106,  96,  86,  76,  66,
      147, 137, 127, 117, 107,  97,  87,  77,
      158, 148, 138, 128, 118, 108,  98,  88,
   );
}

sub print_player_victory {
   print <<'END';

C O N G R A T U L A T I O N S . . .

YOU HAVE WON--VERY WELL PLAYED.
IT LOOKS LIKE I HAVE MET MY MATCH.
THANKS FOR PLAYING---I CAN'T WIN ALL THE TIME.

END
}

sub print_computer_victory {
   print <<'END';

NICE TRY, BUT IT LOOKS LIKE I HAVE WON.
THANKS FOR PLAYING.

END
}

sub forfeit { say "\nIT LOOKS LIKE I HAVE WON BY FORFEIT.\n" }

sub ask_input ($prompt) {
   print "$prompt? ";
   defined(my $input = <STDIN>) or __exit();

   # remove spaces from the input (including newlines), they are not used
   $input =~ s{\s+}{}gmxs;

   return $input;
}

sub ask_yes_no ($prompt) {
   while ('necessary') {
      my $input = ask_input($prompt);
      return 1 if $input =~ m{\A (?: yes | y) \z}imxs;
      return 0 if $input =~ m{\A (?:  no | n) \z}imxs;
      say q{PLEASE ANSWER 'YES' OR 'NO'.};
   }
}

sub ask_first_move {
   while ('necessary') {
      my $input = ask_input('WHERE WOULD YOU LIKE TO START');
      if ($input =~ m{\A (?: 0 | [1-9]\d+) \z}mxs) {
         return 0 unless $input;
         my $diagonal = int($input / 10);
         my $row = $input % 10;
         return $input if $row == 1 || $row == $diagonal;
      }
      say <<'END'
PLEASE READ THE DIRECTIONS AGAIN.
YOU HAVE BEGUN ILLEGALLY.

END
   }
}

sub __exit {
   say "\nOK --- THANKS AGAIN.";
   exit 0;
}

sub welcome {
   print <<'END'
                                 QUEEN
               CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY



END
}

sub help {
   print <<'END';
WE ARE GOING TO PLAY A GAME BASED ON ONE OF THE CHESS
MOVES.  OUR QUEEN WILL BE ABLE TO MOVE ONLY TO THE LEFT,
DOWN, OR DIAGONALLY DOWN AND TO THE LEFT.

THE OBJECT OF THE GAME IS TO PLACE THE QUEEN IN THE LOWER
LEFT HAND SQUARE BY ALTERNATING MOVES BETWEEN YOU AND THE
COMPUTER.  THE FIRST ONE TO PLACE THE QUEEN THERE WINS.

YOU GO FIRST AND PLACE THE QUEEN IN ANY ONE OF THE SQUARES
ON THE TOP ROW OR RIGHT HAND COLUMN.
THAT WILL BE YOUR FIRST MOVE.
WE ALTERNATE MOVES.
YOU MAY FORFEIT BY TYPING '0' AS YOUR MOVE.
BE SURE TO PRESS THE RETURN KEY AFTER EACH RESPONSE.


END
}

sub print_board {
   say '';
   my @ids = board_identifiers();
   my $row_template = join '  ', ($ENV{ORIGINAL} ? '%d' : '%3d') x 8;
   for my $A (0 .. 7) {
      my $start = $A * 8;
      my @range = $start .. $start + 7;
      say ' ', sprintf $row_template, @ids[@range];
      say "\n";
   }
   say '';
}

File: 72_Queen/python/queen.py
"""Implementation of queen.bas to Python."""

from typing import Final, Optional
from random import random


########################################################################################
#                                  Optional configs
########################################################################################
# You can edit these variables to change the behavior of the game.
#
# The original implementation has a bug that allows a player to move off the board,
# e.g. start at the nonexistant space 91. Change the variable FIX_BOARD_BUG to ``True``
# to fix this behavior.
#

FIX_BOARD_BUG: Final[bool] = False

# In the original implementation, the board is only printed once. Change the variable
# SHOW_BOARD_ALWAYS to ``True`` to display the board every time.

SHOW_BOARD_ALWAYS: Final[bool] = False

# In the original implementaiton, the board is printed a bit wonky because of the
# differing widths of the numbers. Change the variable ALIGNED_BOARD to ``True`` to
# fix this.

ALIGNED_BOARD: Final[bool] = False

########################################################################################

INSTR_TXT: Final[
    str
] = """WE ARE GOING TO PLAY A GAME BASED ON ONE OF THE CHESS
MOVES.  OUR QUEEN WILL BE ABLE TO MOVE ONLY TO THE LEFT,
DOWN, OR DIAGONALLY DOWN AND TO THE LEFT.

THE OBJECT OF THE GAME IS TO PLACE THE QUEEN IN THE LOWER
LEFT HAND SQUARE BY ALTERNATING MOVES BETWEEN YOU AND THE
COMPUTER.  THE FIRST ONE TO PLACE THE QUEEN THERE WINS.

YOU GO FIRST AND PLACE THE QUEEN IN ANY ONE OF THE SQUARES
ON THE TOP ROW OR RIGHT HAND COLUMN.
THAT WILL BE YOUR FIRST MOVE.
WE ALTERNATE MOVES.
YOU MAY FORFEIT BY TYPING '0' AS YOUR MOVE.
BE SURE TO PRESS THE RETURN KEY AFTER EACH RESPONSE.

"""


WIN_MSG: Final[
    str
] = """C O N G R A T U L A T I O N S . . .

YOU HAVE WON--VERY WELL PLAYED.
IT LOOKS LIKE I HAVE MET MY MATCH.
THANKS FOR PLAYING---I CAN'T WIN ALL THE TIME.

"""

LOSE_MSG: Final[
    str
] = """
NICE TRY, BUT IT LOOKS LIKE I HAVE WON.
THANKS FOR PLAYING.

"""


def loc_to_num(location: tuple[int, int], fix_align: bool = False) -> str:
    """Convert a position given by row, column into a space number."""
    row, col = location
    out_str: str = f"{row + 8 - col}{row + 1}"
    if not fix_align or len(out_str) == 3:
        return out_str
    else:
        return out_str + " "


GAME_BOARD: Final[str] = (
    "\n"
    + "\n\n\n".join(
        "".join(f" {loc_to_num((row, col), ALIGNED_BOARD)} " for col in range(8))
        for row in range(8)
    )
    + "\n\n\n"
)


def num_to_loc(num: int) -> tuple[int, int]:
    """Convert a space number into a position given by row, column."""
    row: int = num % 10 - 1
    col: int = row + 8 - (num - row - 1) // 10
    return row, col


# The win location
WIN_LOC: Final[tuple[int, int]] = (7, 0)

# These are the places (other than the win condition) that the computer will always
# try to move into.
COMPUTER_SAFE_SPOTS: Final[frozenset[tuple[int, int]]] = frozenset(
    [
        (2, 3),
        (4, 5),
        (5, 1),
        (6, 2),
    ]
)

# These are the places that the computer will always try to move into.
COMPUTER_PREF_MOVES: Final[
    frozenset[tuple[int, int]]
] = COMPUTER_SAFE_SPOTS | frozenset([WIN_LOC])

# These are the locations (not including the win location) from which either player can
# force a win (but the computer will always choose one of the COMPUTER_PREF_MOVES).
SAFE_SPOTS: Final[frozenset[tuple[int, int]]] = COMPUTER_SAFE_SPOTS | frozenset(
    [
        (0, 4),
        (3, 7),
    ]
)


def str_with_tab(indent: int, text: str, uppercase: bool = True) -> str:
    """Create a string with ``indent`` spaces followed by ``text``."""
    if uppercase:
        text = text.upper()
    return " " * indent + text


def intro():
    """Print the intro and print instructions if desired."""
    print(str_with_tab(33, "Queen"))
    print(str_with_tab(15, "Creative Computing  Morristown, New Jersey"))
    print("\n" * 2)
    if ask("DO YOU WANT INSTRUCTIONS"):
        print(INSTR_TXT)


def get_move(current_loc: Optional[tuple[int, int]]) -> tuple[int, int]:
    """Get the next move from the player."""
    prompt: str
    player_resp: str
    move_raw: int
    new_row: int
    new_col: int
    if current_loc is None:  # It's the first turn
        prompt = "WHERE WOULD YOU LIKE TO START? "
    else:
        prompt = "WHAT IS YOUR MOVE? "
        row, col = current_loc
    while True:
        player_resp = input(prompt).strip()
        try:
            move_raw = int(player_resp)
            if move_raw == 0:  # Forfeit
                return 8, 8
            new_row, new_col = num_to_loc(move_raw)
            if current_loc is None:
                if (new_row == 0 or new_col == 7) and (
                    not FIX_BOARD_BUG or (new_col >= 0 and new_row < 8)
                ):
                    return new_row, new_col
                else:
                    prompt = (
                        "PLEASE READ THE DIRECTIONS AGAIN.\n"
                        + "YOU HAVE BEGUN ILLEGALLY.\n\n"
                        + "WHERE WOULD YOU LIKE TO START? "
                    )
            else:
                if (
                    (new_row == row and new_col < col)  # move left
                    or (new_col == col and new_row > row)  # move down
                    or (new_row - row == col - new_col)  # move diag left and down
                ) and (not FIX_BOARD_BUG or (new_col >= 0 and new_row < 8)):
                    return new_row, new_col
                else:
                    prompt = "Y O U   C H E A T . . .  TRY AGAIN? "

        except ValueError:
            prompt = "!NUMBER EXPECTED - RETRY INPUT LINE\n? "


def random_computer_move(location: tuple[int, int]) -> tuple[int, int]:
    """Make a random move."""
    row, col = location
    if (z := random()) > 0.6:
        # Move down one space
        return row + 1, col
    elif z > 0.3:
        # Move diagonaly (left and down) one space
        return row + 1, col - 1
    else:
        # Move left one space
        return row, col - 1


def computer_move(location: tuple[int, int]) -> tuple[int, int]:
    """Get the computer's move."""
    # If the player has made an optimal move, then choose a random move
    if location in SAFE_SPOTS:
        return random_computer_move(location)
    # We don't need to implmement the logic of checking for the player's win,
    # because that is checked before this function is called.
    row, col = location
    for k in range(7, 0, -1):
        # If the computer can move left k spaces and end in up in a safe spot or win,
        # do it.
        if (new_loc := (row, col - k)) in COMPUTER_PREF_MOVES:
            return new_loc
        # If the computer can move down k spaces and end up in a safe spot or win, do it.
        if (new_loc := (row + k, col)) in COMPUTER_PREF_MOVES:
            return new_loc
        # If the computer can move diagonally k spaces and end up in a safe spot or win,
        # do it.
        if (new_loc := (row + k, col - k)) in COMPUTER_PREF_MOVES:
            return new_loc
        # As a fallback, do a random move. (NOTE: This shouldn't actally happen--it
        # should always be possible to make an optimal move if the player doesn't play
        # in a location in SAFE_SPOTS.
    return random_computer_move(location)


def main_game() -> None:
    """Execute the main game."""
    game_over: bool = False
    location: Optional[tuple[int, int]] = None  # Indicate it is the first turn
    while not game_over:
        location = get_move(location)
        if location == (8, 8):  # (8, 8) is returned when the player enters 0
            print("\nIT LOOKS LIKE I HAVE WON BY FORFEIT.\n")
            game_over = True
        elif location == WIN_LOC:  # Player wins (in lower left corner)
            print(WIN_MSG)
            game_over = True
        else:
            location = computer_move(location)
            print(f"COMPUTER MOVES TO SQUARE {loc_to_num(location)}")
            if location == WIN_LOC:  # Computer wins (in lower left corner)
                print(LOSE_MSG)
                game_over = True
        # The default behavior is not to show the board each turn, but
        # this can be modified by changing a flag at the start of the file.
        if not game_over and SHOW_BOARD_ALWAYS:
            print(GAME_BOARD)


def ask(prompt: str) -> bool:
    """Ask a yes/no question until user gives an understandable response."""
    inpt: str
    while True:
        # Normalize input to uppercase, no whitespace, then get first character
        inpt = input(prompt + "? ").upper().strip()[0]
        print()
        if inpt == "Y":
            return True
        elif inpt == "N":
            return False
        print("PLEASE ANSWER 'YES' OR 'NO'.")
    return False


if __name__ == "__main__":
    intro()
    still_playing: bool = True
    while still_playing:
        print(GAME_BOARD)
        main_game()
        still_playing = ask("ANYONE ELSE CARE TO TRY")
    print("\nOK --- THANKS AGAIN.")
